packages/app/src/cli/services/environment/id-matching.ts

```diff
@@ -1,137 +1,159 @@
-import {MatchingError, RemoteRegistration} from './identifiers.js'
-import {ExtensionRegistration} from '../dev/create-extension.js'
+import {LocalSource} from './identifiers'
+import {MatchingError, RemoteSource} from './identifiers.js'
 import {IdentifiersExtensions} from '../../models/app/identifiers.js'
-import {ExtensionTypes} from '../../constants.js'
 import {err, ok, Result} from '@shopify/cli-kit/common/result'
-import {error, string} from '@shopify/cli-kit'
+import {string} from '@shopify/cli-kit'
+import {difference, partition, pickBy, uniqBy, groupBy} from 'lodash-es'
 
 export interface MatchResult {
   identifiers: IdentifiersExtensions
-  pendingConfirmation: {extension: LocalExtension; registration: ExtensionRegistration}[]
-  toCreate: LocalExtension[]
-  toManualMatch: {local: LocalExtension[]; remote: ExtensionRegistration[]}
+  toConfirm: {local: LocalSource; remote: RemoteSource}[]
+  toCreate: LocalSource[]
+  toManualMatch: {local: LocalSource[]; remote: RemoteSource[]}
 }
 
-export interface LocalExtension {
-  localIdentifier: string
-  graphQLType: string
-  type: ExtensionTypes
-  configuration: {name: string}
+/**
+ * Filter function to match a local and a remote source by type and name
+ */
+const sameTypeAndName = (local: LocalSource, remote: RemoteSource) => {
+  return remote.type === local.graphQLType && string.slugify(remote.title) === string.slugify(local.configuration.name)
 }
 
-export async function automaticMatchmaking(
-  localExtensions: LocalExtension[],
-  remoteRegistrations: ExtensionRegistration[],
-  identifiers: {[localIdentifier: string]: string},
-  registrationIdField: 'id' | 'uuid',
-): Promise<Result<MatchResult, MatchingError>> {
-  if (remoteRegistrations.length > localExtensions.length) {
-    return err('invalid-environment')
-  }
-
-  const validIdentifiers = identifiers
-
-  // Get the local UUID of an extension, if exists
-  const localId = (extension: LocalExtension) => validIdentifiers[extension.localIdentifier]
+/**
+ * Automatically match local and remote sources if they have the same type and name
+ *
+ * If multiple local or remote sources have the same type and name, they can't be matched automatically
+ */
+function matchByNameAndType(
+  local: LocalSource[],
+  remote: RemoteSource[],
+  remoteIdField: 'id' | 'uuid',
+): {matched: IdentifiersExtensions; pending: {local: LocalSource[]; remote: RemoteSource[]}} {
```

Comment:
no reason, when working on it didn't seem very obvious.
How would you name them?
```
type PossibleMatch = {local: LocalSource; remote: RemoteSource}
type ManualMatches = {local: LocalSource[]; remote: RemoteSource[]}
```
??

Reply:
Would it be strange to name them `ExtensionMatch` and `ExtensionMatches`? Or `Match` and `Matches`?

Reply:
umm,
the first one is a Match that the CLI is not sure is right so the user needs to confirm it.
the second one is not an array of matches, is two arrays that the user needs to try to manually match (but maybe is not possible to do it)

@Arkham what do you think?

Reply:
Discussed with Ju offline, couldn't think of good names to represent this (seems easier to ready like it is now) and since this used in limited places and is easy to read we decided to leave it as it is now. Thanks for the feedback anyway @matteodepalo 🙇